### ISSUE TITLE
Kitti3 could support other terminals, applications

### ISSUE THREAD

**Issue by LandingEllipse** [LandingEllipse/kitti3#17]
There's nothing inherently stopping Kitti3 from managing pretty much any application/container which is able to set a telltale - WM_INSTANCE attribute (name or class), title, or app_id (sway).

Rough approach:
- keep "database" of common/known programs and their way of setting a telltale (e.g. Kitty: `--class` for app_id and `--name` for window_instance)
- add new arg: `-p, --program` (default to Kitty for backwards compatibility)
- if PROGRAM is single word and is in DB, use DB entry when spawning
- elif PROGRAM contains fmt placeholder, put NAME in placeholder when spawning
- else either error out(?) or assume user supplies instance; disable spawning
- potentially add new arg: `-i, --identity` (default window_instance/app_id based on WM); sets which criteria to use when searching for NAME

**Comment by LandingEllipse** [comment]
Initial testing with Alacritty has been a bit of a nightmare. Basically, sway's atomic transactions are problematic when an application's resize confirmation leads to a "correct the state to reflect reality" instruction *after* a subsequent resize instruction has landed in the same transaction... 

Specifically, when going from tiled to floating (with client, view and node all referring to aspects of the Alacritty terminal instance we're interested in):
- User toggles floating; `cmd_floating` sets default geometry for the view via `container_floating_set_default_size` and `container_floating_resize_and_center`
- IPC `window::floating` event is sent
- Bind cleanup forces `transaction_commit_dirty`
- Transaction instruction is created for the node (i.e. the view's resize to default geometry)
  - Transaction is immediately committed; the view is told to configure itself (i.e. change its surface size) via `view_configure`
- Kitti3 responds to `window::floating` after <5ms with `resize set` command; `cmd_resize_set` sets desired geometry for the view via `cmd_resize_set_floating`
- IPC cleanup forces `transaction_commit_dirty`
- Transaction instruction is created for the node (i.e. the view's resize to desired geometry)
  - Transaction is *not* immediately committed (because the client is configuring and thus the previous transaction has not been applied?)
- client reports back via `handle_commit` (xdg_shell.c) callback that it has finished configuring its geometry (~7-17ms after trigger, hence the OoO problem)
- Callback sees that the reported surface geometry != the view's geometry; the view's geometry is updated from the reported configured geometry, *which also force-updates the container's pending content (read: view) geometry*
- Callback issues `transaction_commit_dirty_client`
- Existing transaction instruction for node (resize to desired geometry) is overwritten by state corresponding to the reverted-to-default geometry
- Transaction is committed and applied

### Crude workaround
Delay Kitti3's on_float IPC resize command enough to ensure that the Alacritty configure callback occurs before we update to the desired geometry. This is not ideal, as Alacritty's latency is likely to be fairly dependent on system and runtime factors. But a delay as short as 10ms produces the desired results on my system, it seems...

### Potential Sway-centred solution # 1
Implement a way to avoid calling `container_floating_set_default_size` and `container_floating_resize_and_center` when a (given) container is floated. This drops the initial and superfluous configure altogether. Need to check with i3 to see if there's a case to me made from an "i3-compat" point of view to include a config option or such.

### Potential Sway-centred solution # 2 
Disallow `view_update_size` (from `handle_commit`) if node has pending changes? e.g:
```c
struct sway_container_state *pstate = &view->container->pending;
bool tainted = new_geo.width != (int)pstate->content_width ||
		new_geo.height != (int)pstate->content_height;
```
and then update view's geometry, but don't allow `view_update_size` (side effect: force view's geom to pending client size and trigger transaction instruction), if tainted.

There might be unintended side effects to this, but my impression is that floating functionality doesn't have all its kinks worked out in Sway, so the current approach might just have overlooked this issue. I think the tainted condition above is pretty sound: pending content size is used as the basis for the client configuration resulting in `new_geo`, so if the two don't match we're either dealing with a "tainted" container state, or the client failed to resize. The latter is probably a can of worms, but let's see how it behaves...

**Comment by LandingEllipse** [comment]
The delay workaround above seems to work well in practice. I've exposed the `--crosstalk-delay` cli argument so people can adjust if required.

If I can find the time in the future I might try to pursue a sway-centred solution, but as the move reparenting issue hasn't received any feedback yet I'm not sure there is much interest upstream in messing with these aspects of the internals.

Anyway, I consider this issue resolved by 049f16c41023ed47b0c4c5e52ca12991d6735c52, 01246c8fc68d94ccb0293b4dfc8c78e02999d3dd and 9100c0cb9d9b97ab906daf27d748e139c284df8f.
